### PR TITLE
FEATURE: merge images from child to configurable

### PR DESCRIPTION
**Pull request by sylvainraye**

### CODE REVIEW

File: Entity/AkeneoSettings.php [hunk 1]
```diff
@@ -141,7 +141,28 @@ class AkeneoSettings extends Transport
...
+
+    /**
+     * @var boolean
+     *
+     * @ORM\Column(name="akeneo_merge_image_to_parent", type="boolean", options={"default":"false"})
+     */
+    private $akeneoMergeImageToParent = false;
+
+    /**
+     * @var ParameterBagisAkeneoMergeImageToParent
```
Comment (dxops) [inline Entity/AkeneoSettings.php]: Typo

File: Entity/AkeneoSettings.php [hunk 2]
```diff
@@ -506,6 +528,46 @@ public function setAkeneoLocalesList(array $akeneoLocalesList = null)
...
+    {
+        return $this->akeneoAttributesImageList;
+    }
+
+    /**
+     * @param string $akeneoAttributesImageList
+     *
+     * @return $this
+     */
+    public function setAkeneoAttributesImageList($akeneoAttributesImageList)
```
Comment (dxops) [inline Entity/AkeneoSettings.php]: Add typehint and return type please

File: Entity/AkeneoSettings.php [hunk 3]
```diff
@@ -506,6 +528,46 @@ public function setAkeneoLocalesList(array $akeneoLocalesList = null)
...
+    {
+        $this->akeneoAttributesList = $attributeList;
+
+        return $this;
+    }
+
+    /**
+     * @return string|null
+     */
+    public function getAkeneoAttributesImageList(): ?string
```
Comment (dxops) [inline Entity/AkeneoSettings.php:526]: there is no way to make $this->akeneoAttributesImageList null with typehinted setAkeneoAttributesImageList

File: Form/Type/AkeneoSettingsType.php [hunk 4]
```diff
@@ -404,13 +433,16 @@ public function onPreSubmit(FormEvent $event)
...
-                    'choices' => $currencies,
+                    'required'          => false,
+                    'label'             => 'oro.akeneo.integration.settings.akeneo_currencies.label',
+                    'multiple'          => true,
+                    'choices'           => $currencies,
                 ]
             );
 
+            $transportData['akeneoAttributesList'] = str_replace(' ', '', $transportData['akeneoAttributesList']);
+            $transportData['akeneoAttributesImageList'] = str_replace(' ', '', $transportData['akeneoAttributesImageList']);
```
Comment (dxops) [inline Form/Type/AkeneoSettingsType.php]: Requires validation like line above, see my comments in #22

File: ImportExport/Reader/ProductImageReader.php [hunk 5]
```diff
@@ -2,31 +2,76 @@
...
+
+    /**
+     * @var array
+     */
+    private $proceededParent = [];
+
+    /**
+     * @var array|null
+     */
+    private $attributesImageFilter = null;
```
Comment (dxops) [inline ImportExport/Reader/ProductImageReader.php]: Let's make it an array by default

File: ImportExport/Reader/ProductImageReader.php [hunk 6]
```diff
@@ -2,31 +2,76 @@
...
         $images = [];
         foreach ($items as &$item) {
-            foreach ($item['values'] as &$values) {
-                foreach ($values as $value) {
-                    if ('pim_catalog_image' !== $value['type'] || empty($value['data'])) {
-                        continue;
-                    }
+            foreach ($item['values'] as $code => &$values) {
+
+                if (null === $this->attributesImageFilter || in_array($code, $this->attributesImageFilter)) {
```
Comment (dxops) [inline ImportExport/Reader/ProductImageReader.php]: @sylvainraye what's the reason to add a separate field to filter image attributes?

File: Integration/AkeneoTransport.php [hunk 7]
```diff
@@ -164,11 +165,40 @@ public function getChannels()
     /**
      * @param int $pageSize
      *
-     * @return \Iterator
+     * @return \Oro\Bundle\AkeneoBundle\Integration\Iterator\CategoryIterator
      */
     public function getCategories(int $pageSize)
     {
-        return $this->client->getCategoryApi()->all($pageSize);
+        $categoryTreeChannel = null;
```
Comment (dxops) [inline Integration/AkeneoTransport.php]: Create a separate PR for this fix please

File: ImportExport/Reader/ProductImageReader.php [hunk 8]
```diff
@@ -2,31 +2,76 @@
...
-                    $images[] = [
-                        'SKU' => $item['identifier'] ?? $item['code'],
-                        'Name' => basename($value['data']),
-                    ];
+                        if ($this->getTransport()->isAkeneoMergeImageToParent()
+                            && !empty($item['parent'])
+                            && !isset($this->proceededParent[$identifier])
+                        ) {
+
+                            $this->proceededParent[$identifier] = true;
```
Comment (dxops) [inline ImportExport/Reader/ProductImageReader.php]: Can you add some description how to get the same identifier imported twice here?

File: Resources/translations/messages.en.yml [hunk 9]
```diff
@@ -38,10 +38,19 @@ oro:
...
+        akeneo_attribute_list:
+          label: 'Attribute Filter'
+          tooltip: 'This field enables you to apply filters to sync only the attributes you want. Values must be attribute code, separated with a semi-colon. IMPORTANT: if not defined before to save the integration, all attributes will be imported.'
+        akeneo_attribute_image_list:
+          label: 'Image Attributes Filter'
+          tooltip: 'List of image attributes of Akeneo to use as main product picture.'
         price_list:
           label: 'Pricelist'
+        merge_image:
+          label: 'Merge images from children products to parent'
```
Comment (dxops) [inline Resources/translations/messages.en.yml]: Update this description please
children=simple
parent=configurable

File: Resources/translations/messages.en.yml [hunk 10]
```diff
@@ -38,10 +38,19 @@ oro:
         akeneo_product_filter:
           label: 'Product Filter'
           tooltip: 'This field enables you to apply filters to sync only the products you want. As this filter is passed via API request, it must be filled in JSON format. Details on the format and filter options available for the products can be found in the <a target="_blank" href="https://api.akeneo.com/documentation/filter.html#filters">Filters section of the Akeneo PIM documentation</a>'
+        akeneo_attribute_list:
+          label: 'Attribute Filter'
+          tooltip: 'This field enables you to apply filters to sync only the attributes you want. Values must be attribute code, separated with a semi-colon. IMPORTANT: if not defined before to save the integration, all attributes will be imported.'
+        akeneo_attribute_image_list:
+          label: 'Image Attributes Filter'
+          tooltip: 'List of image attributes of Akeneo to use as main product picture.'
```
Comment (dxops) [inline Resources/translations/messages.en.yml]: Separator and validation descriptions are missing

File: ImportExport/Reader/ProductImageReader.php (hunk 6 above)
Comment (dxops) [reply ImportExport/Reader/ProductImageReader.php]: Not sure that `List of image attributes of Akeneo to use as main product picture.` matches the current behavior.

This condition will skip images, but what if we have three different images for different purposes?
Main image configured in https://github.com/oroinc/OroAkeneoBundle/blob/3.1/ImportExport/DataConverter/ProductImageDataConverter.php#L15 and https://github.com/oroinc/OroAkeneoBundle/blob/3.1/ImportExport/Strategy/ProductImageImportStrategy.php#L72

**Review by dxops — CHANGES_REQUESTED** [review]
Hi @sylvainraye ,

thank you for your contributions, check my comments below please

File: ImportExport/Reader/ProductImageReader.php (hunk 6 above)
Comment (sylvainraye) [reply ImportExport/Reader/ProductImageReader.php]: `what's the reason to add a separate field to filter image attributes?`
-> In akeneo you can have many pictures for different use cases: for print, for web, for anything else. In product, those image may be at product or product model level. Currently, the connector import all pictures of a product and set them as product image which is totally a non wanted behavior and may conduct to unexpected behavior. Example, a customer may have a shop running on Magento and on OroCommerce and may need to specify three different attributes for the different sizes in Magento, while in OroCommerce only one attribute could be used. It's also relevant for project migration (for one shop system to an other), well there are many use cases indeed. Here, we give the flexibility to set which attributes should be defined as product image. 

`Not sure that List of image attributes of Akeneo to use as main product picture. matches the current behavior.`the term `main` picture is probably not appropriate, the logic here set images as Product images. It doesn't affect if an image will be a main image or a small one or else. I changed the wording. The current logic of your connector is that it takes the first image it found and set it as main product image. Which is also open for discussion regarding this logic. Hw it's an other topic that the PR doesn't cover.

**Comment by dxops** [comment]
Hello, @sylvainraye , form updates are missing, so it's not possible to create an integration.

Can we merge this PR with #22 also, to decrease code review and QA time?

File: Entity/AkeneoSettings.php (hunk 3 above)
Comment (sylvainraye) [reply Entity/AkeneoSettings.php:526]: fixed

**Comment by sylvainraye** [comment]
Hi @dxops yes you can merge the PR with #22

**Comment by dxops** [comment]
Fixed by https://github.com/oroinc/OroAkeneoBundle/commit/2a654776c0bbafb1fffc8059686fe9c8c9bf7ee7

**Pull request closed without merging**